### ISSUE TITLE
How to propagate context value changes to delayslot instructions?

### ISSUE THREAD

**Issue by JSquared-dev** [NationalSecurityAgency/ghidra#4259]
I'm trying to implement an instruction with a delayslot that has some effect on the instruction in the delayslot. The first instruction provides some immediate value to the delayslot instruction to expand its shorter immediate field. However, setting a context register with this immediate value does not get propagated to the delayslot instruction. Is this correct behaviour? and if so, is my desired functionality possible?

The documentation for delayslots states that `inst_next` will have a slightly different meaning when a delayslot is used - in the disassembly action section it will refer to the address after the first instruction, and in the semantic section it will refer to the address after the final delayslot. Therefore, using `globalset(inst_next, extended_imm);` in the first instruction should set the `extended_imm` context register for the delayslot, correct? However, the behaviour I am seeing is that the delayslot has the exact same context as the beginning of the first instruction, wherein `extended_imm` is set to 0.

I've verified my SLEIGH implementation by using the modified context register value in the first instruction's semantic section to demonstrate the value has changed as desired, but the delayslot instruction will only receive a 0 value.

Furthermore, after disassembling the instruction pair, I can right click the delayslot instruction in the disassembly listing view and modify the Processor Options to manually set the `extended_imm` value. Then re-disassembling the instruction pair results in the desired behaviour - the delayslot instruction has the extended immediate value and disassembles correctly. So it seems like the desired behaviour is possible, but not directly from SLEIGH?

```
define context contextreg
    extended_imm = (0,25)
;

# inst_next here should refer to the loadri instruction
:extimm extimm_u26 is opcode=0x30 & extimm_u26 [ extended_imm = (extimm_u26 << 0); globalset(inst_next, extended_imm); ] { delayslot(4); }

# extended_imm is now 0 and not what was set above because we're in the delayslot?
loadri_u32: reloc is loadri_u6 [ reloc = (extended_imm << 6) + loadri_u6; ] { export *[const]:4 reloc; }

:loadri loadri_Rd loadri_u32 is opcode=0x54 & ... & loadri_Rd & loadri_u32 {
    loadri_Rd = loadri_u32;
}
```

**Comment by ghidra1** [comment]
- I assume your context spec above is incomplete - could you provide complete contextreg spec?  
- Are you able to provide your full language spec and some sample bytes to demonstrate the issue?  
- Are your instructions fixed-length?  If so, have you tried computing the delayslot address instead of using inst_next?  You can also enable display of all register settings in listing via tool options (see  Listing Fields->Register Field) and examine context register settings indicated in listing to see if the context setting is getting applied to the correct address.

**Comment by JSquared-dev** [comment]
As requested, here is a complete (working) language spec that demonstrates the problem.

```
define endian=little;
define alignment=4;

define space ram      type=ram_space      size=4  default;
define space register type=register_space size=4;

define register offset=0x00 size=4 [
	R0 R1 R2 R3 R4 R5 R6 R7
	SP PC
];

define register offset=0x100 size=32 contextreg;
define context contextreg
	immediate_extension = (0, 31)
	next_instr = (32, 63)
;

define token INSTR_TOKEN(32)
	opcode = (0, 7)
	ext_immediate = (8, 31)
	load_Rd = (8, 10)
	load_immediate = (11, 18)
;

attach variables [
	load_Rd
] [
	R0 R1 R2 R3 R4 R5 R6 R7
];

ext_imm: reloc is ext_immediate [ reloc = ext_immediate << 8; immediate_extension = reloc; next_instr = inst_start + 4; globalset(next_instr, immediate_extension); ] { export *[const]:4 reloc; }

:ext ext_imm is opcode = 0x30 & ext_imm {
	delayslot(4);
}

load_immediate_s32: reloc is load_immediate [ reloc = load_immediate + immediate_extension; ] { export *[const]:4 reloc; }

:load load_Rd, load_immediate_s32 is opcode = 0x20 & load_Rd & load_immediate_s32  {
	load_Rd = load_immediate_s32;
}
```

Loading a binary file with the bytes `\x30\xaa\xaa\xaa\x20\xff\x07\x00` will show the behaviour I have been seeing - it should give the instructions `ext 0xaaaaaa00; _load R7, 0xaaaaaaff`. Disassembling the first byte of the sequence appears to set `immediate_extension` correctly (as per your debugging steps, with Register Fields showing), however the second instruction does not use this value. Clearing these instructions and subsequently only disassembling the second instruction works fine (as it's no longer in the delayslot).

In my debugging, I have at least identified that the disassembly actions occur sequentially from the top most level match downwards - in the definition given above, that would be the root table first, then for each subtable in the matched constructor. I think this is counter intuitive as the subtables may cause context to change that could be influence the parent table's disassembly actions. I proved this by using the `immediate_extension` context variable in the root table constructor and its value was always 0 despite being set by the subtable. Therefore I moved the state propagation to the subtable (not ideal) to account for this, but it did not resolve the problem. I think this may be intended behaviour, however, as there are error cases for using subtable symbols in the disassembly action (which this would be a useful case where using the subtable symbol would be good to avoid splitting context changes between operand parsing and the instruction itself).

So, I now have a language spec that sets the `immediate_extension` context register on the instruction in the delayslot, but the value of that context register is not used when decoding the instruction in the delayslot. Is my desired functionality possible, or is there a better way to achieve this?

**Comment by JSquared-dev** [comment]
I've been able to recreate your results. The original bytes example I gave with only 2 instructions appears to never work with your changes - the `_load` instruction will never take the context reg value.

Taking your extended example, I've set noflow on `immediate_extension` and `globalset` using `inst_next`; I've then duplicated the bytes multiple times and it appears to be doing what I want/expect up until the very last delayslot pair.

```
/// trimmed for brevity
        00000060 30 aa aa aa     ext        0xaaaaaa00
             assume immediate_extension = 0xaaaaaa00
        00000064 20 ff 0f 00     _load      R7,0xaaaaaaff
             assume immediate_extension = <UNKNOWN>
        00000068 20 f8 07 00     load       R0,0xff
        0000006c 30 aa bb cc     ext        0xccbbaa00
             assume immediate_extension = 0xccbbaa00
        00000070 20 f9 07 00     _load      R1,0xff
             assume immediate_extension = <UNKNOWN>
        00000074 20 fa 07 00     load       R2,0xff
```

All the earlier instructions disassemble perfectly up until address `00000070` which is not using the context value - could this be because `00000074` is the final instruction in the disassembly and some off-by-one type error occurs in the disassembler? This bug also happens when the disassembler comes across an invalid instruction - the last valid instruction is incorrect, but all preceding instructions are correct. This is the situation I was in originally when discovering this issue.

I'm not familiar with this area of the code base, so unsure where to start to debug. But it would explain why adding more instructions to the example extends the length of correct disassembly.
